Sources/NIOTransportServices/NIOTSConnectionChannel.swift

```diff
@@ -143,9 +114,83 @@ private struct BackpressureManager {
     }
 }
 
+/// Channel options for the connection channel.
+@available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *)
+internal struct TransportServicesOptions {
+    /// Whether autoRead is enabled for this channel.
+    internal var autoRead: Bool = true
+
+    /// Whether we support remote half closure. If not true, remote half closure will
+    /// cause connection drops.
+    internal var supportRemoteHalfClosure: Bool = false
+
+    /// Whether this channel should wait for the connection to become active.
+    internal var waitForActivity: Bool = true
+}
 
 @available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *)
-internal final class NIOTSConnectionChannel {
+internal final class NIOTSConnectionChannel: StateManagedNWConnectionChannel {
+    var parameters: NWParameters {
+        return NWParameters(tls: self.tlsOptions, tcp: self.tcpOptions)
+    }
+    
+    typealias ActiveSubstate = TCPSubstate
+
+    /// A TCP connection may be fully open or partially open. In the fully open state, both
+    /// peers may send data. In the partially open states, only one of the two peers may send
+    /// data.
+    ///
+    /// We keep track of this to manage the half-closure state of the TCP connection.
+    enum TCPSubstate: NWConnectionSubstate {
+        /// Both peers may send.
+        case open
+
+        /// This end of the connection has sent a FIN. We may only receive data.
+        case halfClosedLocal
+
+        /// The remote peer has sent a FIN. We may still send data, but cannot expect to
+        /// receive more.
+        case halfClosedRemote
+
+        /// The channel is "active", but there can be no forward momentum here. The only valid
+        /// thing to do in this state is drop the channel.
+        case closed
+
+        init() {
+            self = .open
+        }
+        
+        /// Close the input side of the TCP state machine.
+        static func closeInput(state: inout ChannelState<NIOTSConnectionChannel.TCPSubstate>) throws {
```

Comment:
Why are these `static`?

Reply:
Because the logic is defined for UDP and TCP separately, in their own enum. However, both enums are wrapped in `ChannelState<Self>`, and this function needs to be aware of the `ChannelState` enum. Since you cannot have overlapping mutating access to `self`, it's not like you can mutate `NIOTSConnectionChannel.TCPSubstate` from within an access to `ChannelState` while providing the parent enum as context.